File: xion-android-demo/app/src/main/java/com/burnt/xiondemo/ui/screens/wallet/WalletViewModel.kt
package com.burnt.xiondemo.ui.screens.wallet

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.burnt.xiondemo.data.model.TransactionResult
import com.burnt.xiondemo.data.model.WalletState
import com.burnt.xiondemo.data.repository.XionRepository
import com.burnt.xiondemo.security.SecureStorage
import com.burnt.xiondemo.util.Constants
import com.burnt.xiondemo.util.Result
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class WalletUiState(
    val address: String? = null,
    val sessionAddress: String? = null,
    val connectionType: String = "Unknown",
    val grantsActive: Boolean = true,
    val balance: String? = null,
    val sbcBalance: String? = null,
    val vaultBalance: String? = null,
    val isBalanceLoading: Boolean = false,
    val blockHeight: Long? = null,
    val chainId: String = Constants.CHAIN_ID,
    val error: String? = null,
    val sessionExpiryWarning: Boolean = false,
    val isDisconnected: Boolean = false,
    val transactions: List<TransactionResult> = emptyList(),
    val bankLinked: Boolean = false
)

@HiltViewModel
class WalletViewModel @Inject constructor(
    private val repository: XionRepository,
    private val secureStorage: SecureStorage
) : ViewModel() {

    private val _uiState = MutableStateFlow(WalletUiState())
    val uiState: StateFlow<WalletUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            repository.walletState.collect { state ->
                when (state) {
                    is WalletState.Connected -> {
                        val previousAddress = _uiState.value.address
                        _uiState.update {
                            it.copy(
                                address = state.metaAccountAddress,
                                sessionAddress = state.sessionKeyAddress,
                                connectionType = "Meta Account",
                                grantsActive = state.grantsActive
                            )
                        }
                        if (previousAddress == null) {
                            loadRecentTransactions()
                        }
                    }
                    is WalletState.Disconnected -> {
                        _uiState.update { it.copy(isDisconnected = true) }
                    }
                    is WalletState.Connecting -> {}
                }
            }
        }

        viewModelScope.launch {
            repository.transactionHistory.collect { inMemoryTxs ->
                if (inMemoryTxs.isNotEmpty()) {
                    // A new tx was just sent — refresh from chain to get full details
                    loadRecentTransactions()
                }
            }
        }

        // Periodic session expiry check (every 60s)
        viewModelScope.launch {
            while (true) {
                delay(60_000)
                checkSessionExpiry()
            }
        }

        refresh()
    }

    fun refresh() {
        loadBalance()
        loadSbcBalance()
        loadVaultBalance()
        loadBlockHeight()
        loadRecentTransactions()
        checkBankLinked()
    }

    private fun checkBankLinked() {
        val bankAddressId = secureStorage.getString(Constants.PREF_BRALE_BANK_ADDRESS_ID)
        _uiState.update { it.copy(bankLinked = !bankAddressId.isNullOrBlank()) }
    }

    private fun loadBalance() {
        viewModelScope.launch {
            _uiState.update { it.copy(isBalanceLoading = true) }
            when (val result = repository.getBalance()) {
                is Result.Success -> {
                    _uiState.update {
                        it.copy(balance = result.data.amount, isBalanceLoading = false)
                    }
                }
                is Result.Error -> {
                    _uiState.update {
                        it.copy(isBalanceLoading = false, error = result.message)
                    }
                }
                is Result.Loading -> {}
            }
        }
    }

    private fun loadSbcBalance() {
        viewModelScope.launch {
            when (val result = repository.getSbcBalance()) {
                is Result.Success -> {
                    _uiState.update { it.copy(sbcBalance = result.data.amount) }
                }
                is Result.Error -> {} // Non-critical
                is Result.Loading -> {}
            }
        }
    }

    private fun loadVaultBalance() {
        viewModelScope.launch {
            when (val result = repository.getVaultBalance()) {
                is Result.Success -> {
                    _uiState.update { it.copy(vaultBalance = result.data.amount) }
                }
                is Result.Error -> {} // Non-critical
                is Result.Loading -> {}
            }
        }
    }

    private fun loadBlockHeight() {
        viewModelScope.launch {
            when (val result = repository.getBlockHeight()) {
                is Result.Success -> {
                    _uiState.update { it.copy(blockHeight = result.data) }
                }
                is Result.Error -> {}
                is Result.Loading -> {}
            }
        }
    }

    private fun loadRecentTransactions() {
        viewModelScope.launch {
            val address = _uiState.value.address ?: return@launch
            val onChainTxs = when (val result = repository.getRecentTransactions(address)) {
                is Result.Success -> result.data
                else -> emptyList()
            }
            // Merge on-chain transactions with in-memory Brale transactions (Buy SBC, Cash Out)
            val inMemoryTxs = repository.transactionHistory.value
            val seen = mutableMapOf<String, TransactionResult>()
            // On-chain first (has enriched data)
            for (tx in onChainTxs) { seen[tx.txHash] = tx }
            // In-memory adds new entries (Brale txs have unique IDs)
            for (tx in inMemoryTxs) { if (tx.txHash !in seen) seen[tx.txHash] = tx }
            val merged = seen.values.sortedByDescending {
                if (it.height > 0) it.height else Long.MAX_VALUE // Brale entries (height=0) sort to top
            }
            _uiState.update { it.copy(transactions = merged) }
        }
    }

    private fun checkSessionExpiry() {
        val state = repository.walletState.value
        if (state is WalletState.Connected) {
            val now = System.currentTimeMillis() / 1000
            val remaining = state.sessionExpiresAt - now
            when {
                remaining <= 0 -> {
                    repository.disconnect()
                }
                remaining < 300 -> {
                    _uiState.update { it.copy(sessionExpiryWarning = true) }
                }
            }
        }
    }

    fun disconnect() {
        repository.disconnect()
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: xion-ios/XionDemo/UI/Screens/Wallet/WalletViewModel.swift
import Combine
import Foundation

@MainActor
final class WalletViewModel: ObservableObject {

    @Published var address: String?
    @Published var sessionAddress: String?
    @Published var connectionType = "Unknown"
    @Published var grantsActive = true
    @Published var balance: String?
    @Published var sbcBalance: String?
    @Published var isBalanceLoading = false
    @Published var blockHeight: Int64?
    @Published var chainId = Constants.chainId
    @Published var error: String?
    @Published var sessionExpiryWarning = false
    @Published var isDisconnected = false
    @Published var transactions: [TransactionResult] = []
    @Published var bankLinked = false

    private let repository: XionRepositoryProtocol
    private let braleRepository: BraleRepositoryProtocol
    private let secureStorage: SecureStorage
    private var cancellables = Set<AnyCancellable>()
    private var expiryTimer: Timer?

    init(
        repository: XionRepositoryProtocol,
        braleRepository: BraleRepositoryProtocol,
        secureStorage: SecureStorage
    ) {
        self.repository = repository
        self.braleRepository = braleRepository
        self.secureStorage = secureStorage

        repository.sessionManager.$walletState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.handleStateChange(state)
            }
            .store(in: &cancellables)

        startExpiryCheck()
        refresh()
    }

    func refresh() {
        checkBankLinked()
        loadBalance()
        loadSbcBalance()
        loadBlockHeight()
        loadTransactions()
    }

    private func checkBankLinked() {
        let bankId = secureStorage.getBraleBankAddressId()
        bankLinked = bankId != nil && !bankId!.isEmpty
    }

    func disconnect() {
        repository.disconnect()
    }

    func clearError() {
        error = nil
    }

    private func loadBalance() {
        Task {
            isBalanceLoading = true
            do {
                let info = try await repository.getBalance()
                balance = info.amount
            } catch {
                self.error = error.localizedDescription
            }
            isBalanceLoading = false
        }
    }

    private func loadSbcBalance() {
        Task {
            do {
                let info = try await repository.getSbcBalance()
                sbcBalance = info.amount
            } catch {
                // Non-critical — SBC balance failure should not show error
            }
        }
    }

    private func loadBlockHeight() {
        Task {
            do {
                blockHeight = try await repository.getBlockHeight()
            } catch {
                // Non-critical
            }
        }
    }

    private func loadTransactions() {
        guard let addr = address else {
            NSLog("[WalletVM] loadTransactions skipped — address is nil")
            return
        }
        Task {
            // Buy/Cash Out come from Brale (they carry pending/processing/complete
            // status, and a pending onramp has no on-chain record yet). Plain XION
            // Send/Received come from on-chain history; drop on-chain SBC legs so an
            // on/offramp isn't shown twice.
            let onChain = (try? await repository.getRecentTransactions(address: addr)) ?? []
            let xionOnly = onChain.filter { !$0.amountDenom.lowercased().contains("sbc") }
            let braleTxs = await loadBraleTransfers()
            let merged = (braleTxs + xionOnly)
                .sorted { $0.timestamp > $1.timestamp }
                .prefix(5)
            transactions = Array(merged)
            NSLog("[WalletVM] %d brale + %d on-chain tx", braleTxs.count, xionOnly.count)
        }
    }

    private func loadBraleTransfers() async -> [TransactionResult] {
        guard let transfers = try? await braleRepository.getRecentTransfers() else { return [] }
        return transfers.compactMap { Self.mapBraleTransfer($0) }
    }

    /// Maps a Brale transfer to a wallet row and surfaces its pending/complete status.
    /// Classify by BOTH legs: an ACH buy pays USD for SBC, a cash out sells SBC for
    /// USD, and an SBC→SBC transfer is someone sending you stablecoin (not a buy).
    private static func mapBraleTransfer(_ t: BraleTransfer) -> TransactionResult? {
        let src = t.source?.valueType.uppercased() ?? ""
        let dst = t.destination?.valueType.uppercased() ?? ""
        let label: String
        switch (src, dst) {
        case ("USD", "SBC"): label = "Buy"       // ACH onramp
        case ("SBC", "USD"): label = "Cash Out"  // ACH offramp
        case ("SBC", "SBC"): label = "Received"  // someone sent you SBC
        default: return nil
        }

        let status: String
        switch t.status.lowercased() {
        case "pending": status = "Pending"
        case "processing": status = "Processing"
        case "complete", "completed": status = "Completed"
        case "canceled", "cancelled": status = "Canceled"
        case "failed": status = "Failed"
        default: status = t.status.capitalized
        }

        let isDone = status == "Completed"
        let isFailed = status == "Failed" || status == "Canceled"
        return TransactionResult(
            txHash: t.id,
            success: !isFailed,
            gasUsed: "0",
            gasWanted: "0",
            height: 0,
            rawLog: "",
            timestamp: t.createdAt ?? "",
            fee: "",
            txType: label,
            amount: "",
            amountDenom: "SBC",
            recipient: "",
            status: status,
            displayAmount: "$\(t.amount.value)",
            inProgress: !isDone && !isFailed
        )
    }

    private func handleStateChange(_ state: WalletState) {
        switch state {
        case .connected(let meta, let session, _, let grants, _):
            address = meta
            sessionAddress = session
            connectionType = "Meta Account"
            grantsActive = grants
            loadTransactions()
        case .disconnected:
            isDisconnected = true
        case .connecting:
            break
        }
    }

    private func startExpiryCheck() {
        expiryTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.checkSessionExpiry()
            }
        }
    }

    private func checkSessionExpiry() {
        let state = repository.sessionManager.walletState
        guard state.isConnected else { return }
        let now = Int64(Date().timeIntervalSince1970)
        let remaining = state.sessionExpiresAt - now
        if remaining <= 0 {
            repository.disconnect()
        } else if remaining < 300 {
            sessionExpiryWarning = true
        }
    }

    deinit {
        expiryTimer?.invalidate()
    }
}
